I get a segfault while first start

Hi, 

First I want to thank for your hard work to make this possible. 

So the problem is, after I compiled everything and wanted to start it for the first time I've got 
an "bodo@donald:~/CTP2/ctp2_code/ctp> ./ctp2
        Segmentation fault (core dumped)" in the terminal. 

I'm running it or tried to run it on an Opensuse 15.6-64 bit machine. 
I have no idea how to fix it. 
What can I do to give you more info about this problem? 

The message above is all I get in the terminal. 

Thx
Bodo

It's me again. After a break I thought I try it again. 

First I see that after autogen.sh I see a message:
 configure.ac:34: warning: AC_OUTPUT should be used without arguments.
configure.ac:34: You should run autoupdate.
But ignored this for now. 

After this I've executed 
CFLAGS="$CFLAGS -O1 -fuse-ld=gold" CXXFLAGS="$CXXFLAGS -fpermissive -O3 -fuse-ld=gold"

with the only change of $CFLAGS -O3 to $CFLAGS -O1

Then i ran ./configure --enable-silent-rules > ./configure-output.txt
and I've got the message with the "WARNING: 'missing' script is too old or missing
sed: -e expression https://github.com/civctp2/civctp2/issues/1, char 1: unknown command: `-' "

but this doesn't show up when I just run ./configure --enable-silent-rules.

I don't know why I get a warning about something missing when I redirect the output from the terminal in a txt.file. 

After this I've executed 
make -j$(nproc) > ./make-output.txt
and I get now this:
/bin/sh: patch: command not found
make[5]: *** [GNUmakefile:479: d3des/unpacked] Error 127
make[5]: *** Waiting for unfinished jobs....
make[4]: *** [GNUmakefile:365: all-recursive] Error 1
make[3]: *** [GNUmakefile:417: all-recursive] Error 1
make[2]: *** [Makefile:415: all-recursive] Error 1
make[1]: *** [Makefile:3269: all-recursive] Error 1
make: *** [GNUmakefile:468: all-recursive] Error 1

I get the same massage when I run make without redirecting to a file. 
Just make -j$(nproc)

OK, I've found the problem but I don't know why it didn't show up when I first compiled it. 
I had patch and patch-utils not installed. 

After I installed the 2 packages I think make finished but at the end I see: 
make[1]: Entering directory '/home/bodo/CTP2'
make[1]: Nothing to be done for 'all-am'.
make[1]: Leaving directory '/home/bodo/CTP2'

What is all-am?

Hi Martin, 

Now I recovered from a surgery and tried your newer version. 
But I have problems to compile it. 

Configure goes through without the  "-fuse-ld=gold " option and 
while running make I get this:

Making all in dbgen
make[3]: Verzeichnis „/home/bodo/CTP2/ctp2_code/gs/dbgen“ wird betreten
  CXXLD    ctpdb
collect2: fatal error: cannot find 'ld'
compilation terminated.
make[3]: *** [Makefile:438: ctpdb] Fehler 1
make[3]: Verzeichnis „/home/bodo/CTP2/ctp2_code/gs/dbgen“ wird verlassen
make[2]: *** [Makefile:422: all-recursive] Fehler 1
make[2]: Verzeichnis „/home/bodo/CTP2/ctp2_code/gs“ wird verlassen
make[1]: *** [Makefile:3279: all-recursive] Fehler 1
make[1]: Verzeichnis „/home/bodo/CTP2/ctp2_code“ wird verlassen
make: *** [GNUmakefile:468: all-recursive] Fehler 1

Then I've added the -fuse-ld=gold cuz I've got the ld problem as you've mentioned in your readme.
I've tried the debug option too but it's the same error.

Now with configure I get this error: 

bodo@donald:~/CTP2> CFLAGS="$CFLAGS -O3 -fuse-ld=gold" CXXFLAGS="$CXXFLAGS -O3 -fuse-ld=gold" ./configure --enable-silent-rules
configure: loading site script /usr/share/site/x86_64-unknown-linux-gnu
checking build system type... x86_64-pc-linux-gnu
checking host system type... x86_64-pc-linux-gnu
checking target system type... x86_64-pc-linux-gnu
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a race-free mkdir -p... /usr/bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking whether to enable maintainer-specific portions of Makefiles... no
checking whether make sets $(MAKE)... (cached) yes
checking for cl... no
checking for gcc... gcc
checking whether the C compiler works... no
configure: error: in '/home/bodo/CTP2':
configure: error: C compiler cannot create executables
See 'config.log' for more details


I will attach the config.log as it might help. 

[config.log](https://github.com/user-attachments/files/24046597/config.log)

So I solved the ld problem by installing binutils.gold. 
autogen and config is running through but while compiling is an error: 
For me it looks like there's an error in "ui/aui_sdl/aui_sdlmovie.cpp"

--snip
ui/aui_sdl/aui_sdlmovie.cpp:1999:6: error: ‘c3files_PathIsValid’ was not declared in this scope
  if(!c3files_PathIsValid(m_filename))
      ^~~~~~~~~~~~~~~~~~~
At global scope:
cc1plus: warning: unrecognized command line option ‘-Wno-class-memaccess’
make[2]: *** [Makefile:3242: ui/aui_sdl/aui_sdlmovie.o] Fehler 1
make[2]: *** Es wird auf noch nicht beendete Prozesse gewartet....
At global scope:
cc1plus: warning: unrecognized command line option ‘-Wno-class-memaccess’
ui/aui_utils/primitives.cpp: In function ‘void primitives_ClippedTriangle16(aui_Surface&, const RECT&, TRIANGLE_ID, Pixel16, uint8, bool)’:
ui/aui_utils/primitives.cpp:49:51: warning: ‘y3’ may be used uninitialized in this function [-Wmaybe-uninitialized]
 #define SWAPVARS(a, b) { sint32 temp; temp = a; a = b; b = temp; }
                                                   ^
ui/aui_utils/primitives.cpp:2970:9: note: ‘y3’ was declared here
  sint32 y3;
         ^~
ui/aui_utils/primitives.cpp:49:51: warning: ‘x3’ may be used uninitialized in this function [-Wmaybe-uninitialized]
 #define SWAPVARS(a, b) { sint32 temp; temp = a; a = b; b = temp; }
                                                   ^
ui/aui_utils/primitives.cpp:2969:9: note: ‘x3’ was declared here
  sint32 x3;
         ^~
At global scope:
cc1plus: warning: unrecognized command line option ‘-Wno-class-memaccess’
make[2]: Verzeichnis „/home/bodo/CTP2/ctp2_code“ wird verlassen
make[1]: *** [Makefile:3279: all-recursive] Fehler 1
make[1]: Verzeichnis „/home/bodo/CTP2/ctp2_code“ wird verlassen
make: *** [GNUmakefile:468: all-recursive] Fehler 1

> It is created by the debug version on start up, if it does not exist. Otherwise, its content is deleted and new log files are created. The crash.txt also goes there if it exists, otherwise it goes to the directory with the executable.
> 
> Actually, I thought you already tried the debug version. Just make sure that you recompile all the code when switching to a debug version and back. To do so just use `make clean`.
> 
I always use make clean before I recompile something. ;-)

Ya, look in my answer just a moment ago. 
I've used:

autogen.sh
CFLAGS="$CFLAGS -g -O0 -fno-omit-frame-pointer" CXXFLAGS="$CXXFLAGS -g -O0 -fno-omit-frame-pointer" ./configure
make -j$(nproc)

> > It was already enabled but there is no folder logs. But there is a crash.txt in /home/bodo/CTP2/ctp2_code/ctp/ but it's empty.
> That's not good. Especially, if it is just empty. There should be some lines that are not part of the stack trace.
no, it's empty. where should the folder logs be? In the main folder or some subfolder?

> > The intro video I found and could play it with vlc but when I started the game it didn't play.
> > The startscreen is empty with no sound. If I click in the empty window I get the start screen for singleplay aso.
> That's how it behaves if it has been compiled without video support. Looks like it did not recognize it. Configure should output:
> checking for libavcodec/avcodec.h... yes
> checking for libavformat/avformat.h... yes
> checking for libswscale/swscale.h... yes

configure:18465: checking for libavcodec/avcodec.h
configure:18465: g++ -c  -g -O0 -fno-omit-frame-pointer  conftest.cpp >&5
conftest.cpp:54:10: fatal error: libavcodec/avcodec.h: No such file or directory

configure:18474: checking for libavformat/avformat.h
configure:18474: g++ -c  -g -O0 -fno-omit-frame-pointer  conftest.cpp >&5
conftest.cpp:54:10: fatal error: libavformat/avformat.h: No such file or directory

configure:18483: checking for libswscale/swscale.h
configure:18483: g++ -c  -g -O0 -fno-omit-frame-pointer  conftest.cpp >&5
conftest.cpp:54:10: fatal error: libswscale/swscale.h: No such file or directory

But I have them in:
/usr/include/ffmpeg/libavcodec/avcodec.h
/usr/include/ffmpeg/libavformat/avformat.h
/usr/include/ffmpeg/libswscale/swscale.h

Name        : ffmpeg-6-libavcodec-devel
Version     : 6.1.3
Release     : 150600.2.pm.2
Architecture: x86_64

Name        : ffmpeg-6-libavformat-devel
Version     : 6.1.3
Release     : 150600.2.pm.2
Architecture: x86_64

Name        : ffmpeg-6-libswscale-devel
Version     : 6.1.3
Release     : 150600.2.pm.2
Architecture: x86_64

all from Packman in the same version. 

I just looked and my gcc is version 7.5.0 but I have 
gcc 11 too but the the call is gcc-11 and it gives me gcc version 11.3.0 (SUSE Linux)

So 
gcc -v -> 7.5.0
gcc-11 -v -> 11.3.0

btw.  pkg-config is 0.29.2

That's my config.log and it shows it's not compiled but the files are present. 
But I've seen that there are some gcc errors where my gcc doesn't know these options. For example 
gcc: error: unrecognized command line option '-V'
gcc: error: unrecognized command line option '-qversion'; did you mean '--version'?
gcc: error: unrecognized command line option '-version'

[config.log](https://github.com/user-attachments/files/24123162/config.log)

--- a/ctp2_code/sound/civsound.cpp
+++ b/ctp2_code/sound/civsound.cpp
@@ -77,14 +77,13 @@ CivSound::CivSound(const uint32 &associatedObject, const sint32 &soundID)
     if (!fname) {
         return;
     }
-#if defined(USE_SDL)
+
 	// "NULL.WAV" contains no audio data, and with this file SDL 2.x
 	// returns a pointer that it has already freed, which it later
 	// tries to free again in Mix_FreeChunk.
 	if (strcasecmp(fname, "NULL.WAV") == 0) {
 		return;
 	}
-#endif
 
     size_t      l_dataSize = 0;
     m_dataptr   = g_SoundPF->getData(fname, l_dataSize);

--- a/ctp2_code/sound/civsound.cpp
+++ b/ctp2_code/sound/civsound.cpp
@@ -59,6 +59,11 @@ CivSound::CivSound(const uint32 &associatedObject, const sint32 &soundID)
     m_associatedObject(associatedObject)
 
 {
+	// Initialize fields in case of early return.
+	m_soundFilename[0] = 0;
+	m_dataptr = NULL;
+	m_datasize = 0;
+
     const char *fname;
 	if(soundID < 0)
 		fname = NULL;
@@ -70,9 +75,6 @@ CivSound::CivSound(const uint32 &associatedObject, const sint32 &soundID)
     m_isLooping = FALSE;
 
     if (!fname) {
-        m_soundFilename[0] = 0;
-        m_dataptr = NULL;
-        m_datasize = 0;
         return;
     }
 #if defined(USE_SDL)
@@ -80,18 +82,18 @@ CivSound::CivSound(const uint32 &associatedObject, const sint32 &soundID)
 	// returns a pointer that it has already freed, which it later
 	// tries to free again in Mix_FreeChunk.
 	if (strcasecmp(fname, "NULL.WAV") == 0) {
-		m_soundFilename[0] = 0;
-		m_dataptr = NULL;
-		m_datasize = 0;
 		return;
 	}
 #endif
 
-    strcpy(m_soundFilename, fname);
-
     size_t      l_dataSize = 0;
-    m_dataptr   = g_SoundPF->getData(m_soundFilename, l_dataSize);
-    m_datasize  = static_cast<sint32>(l_dataSize);
+    m_dataptr   = g_SoundPF->getData(fname, l_dataSize);
+	if (!m_dataptr) {
+		return;
+	}
+
+    strcpy(m_soundFilename, fname);
+    m_datasize = static_cast<sint32>(l_dataSize);
 
 #if !defined(USE_SDL)
 	m_hAudio = AIL_quick_load_mem(m_dataptr, m_datasize);

--- a/ctp2_code/sound/civsound.cpp
+++ b/ctp2_code/sound/civsound.cpp
@@ -87,7 +87,14 @@ CivSound::CivSound(const uint32 &associatedObject, const sint32 &soundID)
     // Argh, audio format mismatch!!!
 	m_Audio = Mix_QuickLoad_RAW((Uint8 *) m_dataptr, (Uint32) m_datasize);
 # else
-    m_Audio = Mix_LoadWAV_RW(SDL_RWFromMem(m_dataptr, static_cast<int>(m_datasize)), 1);
+	m_Audio = Mix_LoadWAV_RW(SDL_RWFromMem(m_dataptr, static_cast<int>(m_datasize)), 1);
+	// Compensate for possible SDL bug: if alen=0, then abuf may have been freed by
+	// realloc inside SDL. Since the behavior of realloc(ptr, 0) is implementation
+	// dependent, conservatively assume that realloc has freed the memory, and set abuf
+	// to NULL to prevent it from being freed again by SDL's Mix_FreeChunk.
+	if (!m_Audio->alen) {
+		m_Audio->abuf = NULL;
+	}
 # endif
 #endif
 }

--- a/ctp2_code/sound/civsound.cpp
+++ b/ctp2_code/sound/civsound.cpp
@@ -36,8 +36,6 @@
 //
 //----------------------------------------------------------------------------
 
-#include <strings.h>
-
 #include "c3.h"
 
 #include "civsound.h"
@@ -78,10 +76,13 @@ CivSound::CivSound(const uint32 &associatedObject, const sint32 &soundID)
         return;
     }
 
-	// "NULL.WAV" contains no audio data, and with this file SDL 2.x
-	// returns a pointer that it has already freed, which it later
-	// tries to free again in Mix_FreeChunk.
-	if (strcasecmp(fname, "NULL.WAV") == 0) {
+	// "NULL.WAV" contains no audio data, and for this file SDL 2.x
+	// returns a pointer that it has already freed. It later tries to
+	// free it again in Mix_FreeChunk.
+	// The file is located inside an archive, and archive look-up is
+	// case-insensitive, however "NULL.WAV" is what appears in the
+	// sound configuration files, so that is what we check here.
+	if (strcmp(fname, "NULL.WAV") == 0) {
 		return;
 	}
 

--- a/ctp2_code/sound/civsound.cpp
+++ b/ctp2_code/sound/civsound.cpp
@@ -36,6 +36,8 @@
 //
 //----------------------------------------------------------------------------
 
+#include <strings.h>
+
 #include "c3.h"
 
 #include "civsound.h"
@@ -59,7 +61,7 @@ CivSound::CivSound(const uint32 &associatedObject, const sint32 &soundID)
 {
     const char *fname;
 	if(soundID < 0)
-		fname = 0;
+		fname = NULL;
 	else
 		fname = g_theSoundDB->Get(soundID)->GetValue();
 
@@ -67,12 +69,23 @@ CivSound::CivSound(const uint32 &associatedObject, const sint32 &soundID)
     m_isPlaying = FALSE;
     m_isLooping = FALSE;
 
-    if (0 == fname) {
+    if (!fname) {
         m_soundFilename[0] = 0;
         m_dataptr = NULL;
         m_datasize = 0;
         return;
     }
+#if defined(USE_SDL)
+	// "NULL.WAV" contains no audio data, and with this file SDL 2.x
+	// returns a pointer that it has already freed, which it later
+	// tries to free again in Mix_FreeChunk.
+	if (strcasecmp(fname, "NULL.WAV") == 0) {
+		m_soundFilename[0] = 0;
+		m_dataptr = NULL;
+		m_datasize = 0;
+		return;
+	}
+#endif
 
     strcpy(m_soundFilename, fname);
 
@@ -88,13 +101,8 @@ CivSound::CivSound(const uint32 &associatedObject, const sint32 &soundID)
 	m_Audio = Mix_QuickLoad_RAW((Uint8 *) m_dataptr, (Uint32) m_datasize);
 # else
 	m_Audio = Mix_LoadWAV_RW(SDL_RWFromMem(m_dataptr, static_cast<int>(m_datasize)), 1);
-	// Compensate for possible SDL bug: if alen=0, then abuf may have been freed by
-	// realloc inside SDL. Since the behavior of realloc(ptr, 0) is implementation
-	// dependent, conservatively assume that realloc has freed the memory, and set abuf
-	// to NULL to prevent it from being freed again by SDL's Mix_FreeChunk.
-	if (!m_Audio->alen) {
-		m_Audio->abuf = NULL;
-	}
+	// If alen=0, then abuf may have been freed by realloc inside SDL 2.x.
+	Assert(m_Audio->alen);
 # endif
 #endif
 }

